cosi/cosi.go

```diff
@@ -1,450 +1,399 @@
-	hash.Write(aggPublicMarshal)
+	hash.Write(VBuff)
+	hash.Write(ABuff)
+	hash.Write(mask.mask)
 	hash.Write(message)
 	buff := hash.Sum(nil)
 	k := suite.Scalar().SetBytes(buff)
 
 	// k * -aggPublic + s * B = k*-A + s*B
 	// from s = k * a + r => s * B = k * a * B + r * B <=> s*B = k*A + r*B
 	// <=> s*B + k*-A = r*B
-	minusPublic := suite.Point().Neg(aggPublic)
+	minusPublic := suite.Point().Neg(A)
 	kA := suite.Point().Mul(minusPublic, k)
-	sB := suite.Point().Mul(nil, sigInt)
+	sB := suite.Point().Mul(nil, r)
 	left := suite.Point().Add(kA, sB)
 
-	if !left.Equal(aggCommit) {
-		return errors.New("Signature invalid")
+	x, err := left.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	y, err := V.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	if subtle.ConstantTimeCompare(x, y) == 0 || !policy.Check(mask) {
+		return errors.New("invalid signature")
 	}
-
 	return nil
 }
 
-// AggregateResponse returns the aggregated response that this cosi has
-// accumulated.
-func (c *CoSi) AggregateResponse() abstract.Scalar {
-	return c.aggregateResponse
+// Mask represents a cosigning participation bitmask.
+type Mask struct {
+	mask            []byte
+	publics         []abstract.Point
+	AggregatePublic abstract.Point
 }
 
-// GetChallenge returns the challenge that were passed down to this cosi.
-func (c *CoSi) GetChallenge() abstract.Scalar {
-	return c.challenge
+// NewMask returns a new participation bitmask for cosigning where all
+// cosigners are disabled by default. If a public key is given it verifies that
+// it is present in the list of keys and sets the corresponding index in the
+// bitmask to 1 (enabled).
+func NewMask(suite abstract.Suite, publics []abstract.Point, myKey abstract.Point) (*Mask, error) {
+	m := &Mask{
+		publics: publics,
+	}
+	m.mask = make([]byte, m.Len())
+	m.AggregatePublic = suite.Point().Null()
+	if myKey != nil {
+		found := false
+		for i, key := range publics {
+			if key.Equal(myKey) {
+				m.SetBit(i, true)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, errors.New("key not found")
+		}
+	}
+	return m, nil
 }
 
-// GetCommitment returns the commitment generated by this CoSi (not aggregated).
-func (c *CoSi) GetCommitment() abstract.Point {
-	return c.commitment
+// Mask returns a copy of the participation bitmask.
+func (m *Mask) Mask() []byte {
+	clone := make([]byte, len(m.mask))
+	copy(clone[:], m.mask)
+	return clone
 }
 
-// GetResponse returns the individual response generated by this CoSi
-func (c *CoSi) GetResponse() abstract.Scalar {
-	return c.response
+// SetMask sets the participation bitmask according to the given byte slice
+// interpreted in little-endian order, i.e., bits 0-7 of byte 0 correspond to
+// cosigners 0-7, bits 0-7 of byte 1 correspond to cosigners 8-15, etc.
+func (m *Mask) SetMask(mask []byte) error {
+	if m.Len() != len(mask) {
+		return fmt.Errorf("mismatching mask lengths")
+	}
+	for i := range m.publics {
+		byt := i >> 3
```

Comment:
Because I need the values of `byt` and `msk` inside the `if` statements as well and `Bit` currently only returns `(bool, error)`. Not sure if it would be good to overload `Bit` to return `byt` and `msk` as well. Actually, I just realised that `Bit` is never used in the code anyways, so maybe we can just remove that? ;)